File: Singly_LinkedList.java
import java.util.Scanner;

public class Singly_LinkedList {

    static class Node {
        String nim;
        String nama;
        Node next;

        Node(String nim, String nama) {
            this.nim = nim;
            this.nama = nama;
            this.next = null;
        }
    }

    static Node head = null;

     static void insertBeginning(String nim, String nama) {
        Node newNode = new Node(nim, nama);
        newNode.next = head;
        head = newNode;
        System.out.println("!!DATA BERHASIL DISIMPAN!!");
    }

    static void insertPos(int pos, String nim, String nama) {

        if (pos < 1){
        System.out.println("!!POSISI TIDAK VALID!!");
            return;
        }

        if (pos == 1) {
            Node newNode = new Node(nim, nama);
            newNode.next = head;
            head = newNode;
            return;
        }

        Node curr = head;

    
        for (int i = 1; i < pos - 1 && curr != null; i++) {
            curr = curr.next;
        }

       
        if (curr == null)
            return;

        Node newNode = new Node(nim, nama);

        newNode.next = curr.next;
        curr.next = newNode;
         System.out.println("!!DATA BERHASIL DISIMPAN!!");
        return;
    }

     static void insertEnd(String nim, String nama) {
      
        Node newNode = new Node(nim, nama);

        if (head == null) {
            head = newNode;
        System.out.println("!!DATA BERHASIL DISIMPAN!!");
            return;
        }

        Node last = head;

        while (last.next != null) {
            last = last.next;
        }

        last.next = newNode;

         System.out.println("!!DATA BERHASIL DISIMPAN!!");

        return;
    }

    static void deleteBeginning() {

        if (head == null){
        System.out.println("!!DAFTAR KOSONG!!");
            return;
        }

        head = head.next;
         System.out.println("!!DATA BERHASIL DIHAPUS!!");

        return;
    }

    static void deletePos(int pos) {
        Node temp = head;

        if (head == null || pos < 1){
        System.out.println("!!DATA TIDAK DITEMUKAN!!");
        return;
        }

        if (pos== 1) {
            head = temp.next;
         System.out.println("!!DATA BERHASIL DIHAPUS!!");
            return;
        }

        Node prev = null;
        for (int i = 1; i < pos && temp != null; i++) {
            prev = temp;
            temp = temp.next;
        }

        if (temp == null) return;

        prev.next = temp.next;
        System.out.println("!!DATA BERHASIL DIHAPUS!!");
    }

    static void deleteEnd() {
     
        if (head == null) {
        System.out.println("!!DAFTAR KOSONG!!");
            return;
        }

        if (head.next == null) {
            head = null;
            return;
        }
        
        Node secondLast = head;
        while (secondLast.next.next != null) {
            secondLast = secondLast.next;
        }

    
        secondLast.next = null;
        System.out.println("!!DATA BERHASIL DIHAPUS!!");
    }

    static void deleteFirstOccurrence(String nim) {

    if (head == null){
    System.out.println("!!DAFTAR KOSONG!!");
    return;
    }

    if (head.nim.equals(nim)) {
        head = head.next;
        return;
    }

    Node curr = head;

    while (curr.next != null && !curr.next.nim.equals(nim)) {
        curr = curr.next;
    }

    if (curr.next != null) {
        curr.next = curr.next.next;
    System.out.println("!!DATA BERHASIL DIHAPUS!!");
    }
    
}

static void show() {

    if (head == null) {
        System.out.println("!!DAFTAR KOSONG!!");
        return;
    }
    System.out.println("==DAFTAR MAHASISWA==");

    Node temp = head;
    int i = 0;

    while (temp != null) {
        
        System.out.println(i + ". " + temp.nim + " | " + temp.nama);
        temp = temp.next;
        i++;
    }
}



    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

    

    while (true) {
        System.out.println();
        System.out.println("==================");
        System.out.println(" MENU PROGRAM");
        System.out.println("==================");
        System.out.println("""
1 Insert Beginning
2 Insert Position
3 Insert End
4 Delete Beginning
5 Delete Position
6 Delete End
7 Delete First Occurrence
8 Show
9 Exit
""");

            System.out.print("Menu: ");
            int choice = sc.nextInt();
            sc.nextLine();

    if(choice == 1){
        System.out.print("NIM: ");
        String nim = sc.nextLine();

        System.out.print("Nama: ");
        String nama = sc.nextLine();

        insertBeginning(nim, nama);
    }

    else if(choice == 2){
        System.out.print("Posisi: ");
        int pos = sc.nextInt();
        sc.nextLine();

        System.out.print("NIM: ");
        String nim = sc.nextLine();

        System.out.print("Nama: ");
        String nama = sc.nextLine();

        insertPos(pos, nim, nama);
    }

    else if(choice == 3){
        System.out.print("NIM: ");
        String nim = sc.nextLine();

        System.out.print("Nama: ");
        String nama = sc.nextLine();

        insertEnd(nim, nama);
    }

    else if(choice == 4){
        deleteBeginning();
    }

    else if(choice == 5){
        System.out.print("Posisi: ");
        int pos = sc.nextInt();
        sc.nextLine();

        deletePos(pos);
    }

    else if(choice == 6){
        deleteEnd();
    }

    else if(choice == 7){
        System.out.print("NIM: ");
        String nim = sc.nextLine();
        deleteFirstOccurrence(nim);
    }

    else if(choice == 8){
        show();
    }

    else if(choice == 9){
        System.out.println("!!PROGRAM DIAKHIRI!!");
        break;
    }

        }
    }

}

File: Singly_LinkedList.py
class Node:
    def __init__(self, nim, nama):
        self.nim = nim
        self.nama = nama
        self.next = None


head = None


def insert_beginning(nim, nama):
    global head
    new_node = Node(nim, nama)
    new_node.next = head
    head = new_node
    print("!!DATA BERHASIL DISIMPAN!!")


def insert_pos(pos, nim, nama):
    global head

    if pos < 1:
        print("!!POSISI TIDAK VALID!!")
        return

    new_node = Node(nim, nama)

    if pos == 1:
        new_node.next = head
        head = new_node
        print("!!DATA BERHASIL DISIMPAN!!")
        return

    curr = head

    for _ in range(pos - 2):
        if curr is None:
            print("POSISI MELEBIHI PANJANG LIST")
            return
        curr = curr.next

    if curr is None:
        print("POSISI MELEBIHI PANJANG LIST")
        return

    new_node.next = curr.next
    curr.next = new_node
    print("!!DATA BERHASIL DISIMPAN!!")


def insert_end(nim, nama):
    global head
    new_node = Node(nim, nama)

    if head is None:
        head = new_node
        print("!!DATA BERHASIL DISIMPAN!!")
        return

    curr = head
    while curr.next:
        curr = curr.next

    curr.next = new_node
    print("!!DATA BERHASIL DISIMPAN!!")


def delete_beginning():
    global head
    if head is None:
        print("!!DAFTAR KOSONG!!")
        return

    head = head.next
    print("!!DATA BERHASIL DIHAPUS!!")


def delete_pos(pos):
    global head

    if head is None or pos < 1:
        print("!!DATA TIDAK DITEMUKAN!!")
        return

    if pos == 1:
        head = head.next
        print("!!DATA BERHASIL DIHAPUS!!")
        return

    curr = head
    prev = None

    for _ in range(pos - 1):
        prev = curr
        curr = curr.next
        if curr is None:
            print("!!DATA TIDAK DITEMUKAN!!")
            return

    prev.next = curr.next
    print("!!DATA BERHASIL DIHAPUS!!")


def delete_end():
    global head

    if head is None:
        print("!!DAFTAR KOSONG!!")
        return

    if head.next is None:
        head = None
        print("!!DATA BERHASIL DIHAPUS!!")
        return

    curr = head
    while curr.next.next:
        curr = curr.next

    curr.next = None
    print("!!DATA BERHASIL DIHAPUS!!")


def delete_first_occurrence(nim):
    global head

    if head is None:
        print("!!DAFTAR KOSONG!!")
        return

    if head.nim == nim:
        head = head.next
        print("!!DATA BERHASIL DIHAPUS!!")
        return

    curr = head

    while curr.next and curr.next.nim != nim:
        curr = curr.next

    if curr.next:
        curr.next = curr.next.next
        print("!!DATA BERHASIL DIHAPUS!!")
    else:
        print("NIM TIDAK DITEMUKAN")


def show():
    if head is None:
        print("!!DAFTAR KOSONG!!")
        return

    print("== DAFTAR MAHASISWA ==")
    temp = head
    i = 0

    while temp:
        print(f"{i}. {temp.nim} | {temp.nama}")
        temp = temp.next
        i += 1


while True:
    print("""
==================
 MENU PROGRAM
==================
1 Insert Beginning
2 Insert Position
3 Insert End
4 Delete Beginning
5 Delete Position
6 Delete End
7 Delete First Occurrence
8 Show
9 Exit
""")

    choice = int(input("Menu: "))

    if choice == 1:
        nim = input("NIM: ")
        nama = input("Nama: ")
        insert_beginning(nim, nama)

    elif choice == 2:
        pos = int(input("Posisi: "))
        nim = input("NIM: ")
        nama = input("Nama: ")
        insert_pos(pos, nim, nama)

    elif choice == 3:
        nim = input("NIM: ")
        nama = input("Nama: ")
        insert_end(nim, nama)

    elif choice == 4:
        delete_beginning()

    elif choice == 5:
        pos = int(input("Posisi: "))
        delete_pos(pos)

    elif choice == 6:
        delete_end()

    elif choice == 7:
        nim = input("NIM: ")
        delete_first_occurrence(nim)

    elif choice == 8:
        show()

    elif choice == 9:
        print("!!PROGRAM DIAKHIRI!!")
        break
